doc/sonic-sub-port-intf-hld.md

```diff
@@ -0,0 +1,483 @@
+    "state" : "ok"
+```
+
+Example:
+```
+PORT_TABLE|Ethernet64.10
+    "state" : "ok"
+```
+```
+INTERFACE_TABLE|Ethernet64.10|192.168.0.1/21
+    "state" : "ok"
+```
+
+## 2.4 SAI
+SAI attributes related to a sub port interface are listed in the Table below.
+
+| SAI attributes                                   | attribute value/type                         |
+|--------------------------------------------------|----------------------------------------------|
+| SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID      | VRF oid                                      |
+| SAI_ROUTER_INTERFACE_ATTR_TYPE                   | SAI_ROUTER_INTERFACE_TYPE_SUB_PORT           |
+| SAI_ROUTER_INTERFACE_ATTR_PORT_ID                | parent physical port or port channel oid     |
+| SAI_ROUTER_INTERFACE_ATTR_VLAN_ID                | VLAN oid                                     |
+| SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS        | MAC address                                  |
+| SAI_ROUTER_INTERFACE_ATTR_MTU                    | mtu size                                     |
+
+### 2.4.1 Create a sub port interface
+```
+sai_attribute_t sub_intf_attrs[6];
+
+sub_intf_attrs[0].id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
+sub_intf_attrs[0].value.oid = vrf_oid;
+
+sub_intf_attrs[1].id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
+sub_intf_attrs[1].value.s32 = SAI_ROUTER_INTERFACE_TYPE_SUB_PORT;
+
+sub_intf_attrs[2].id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+sub_intf_attrs[2].value.oid = parent_port_oid;  /* oid of the parent physical port or port channel */
+
+sub_intf_attrs[3].id = SAI_ROUTER_INTERFACE_ATTR_VLAN_ID;
+sub_intf_attrs[3].value.oid = vlan_oid;
+
+sai_mac_t mac = {0x00, 0xe0, 0xec, 0xc2, 0xad, 0xf1};
+sub_intf_attrs[4].id = SAI_ROUTER_INTERFACE_ATTR_SRC_MAC_ADDRESS;
+memcpy(sub_intf_attrs[4].value.mac, mac, sizeof(sai_mac_t));
+
+sub_intf_attrs[5].id = SAI_ROUTER_INTERFACE_ATTR_MTU;
+sub_intf_attrs[5].value.u32 = 9100;
+
+uint32_t sub_intf_attrs_count = 6;
+sai_status_t status = create_router_interface(&rif_id, switch_oid, sub_intf_attrs_count, sub_intf_attrs);
+```
+
+### 2.4.2 Set a sub port interface mtu
+```
+sai_attribute_t sub_intf_attr;
+sub_intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_MTU;
+sub_intf_attr.value.u32 = 9000;
+
+sai_status_t status = set_router_interface_attribute(rif_id, &attr);
+```
+
+### 2.4.3 Remove a sub port interface
+```
+sai_status_t status = remove_router_interface(rif_id);
+```
+
+## 2.5 Linux integration
+### 2.5.1 Host sub port interfaces
+
+We use iproute2 package to manage host sub port interfaces.
+Specifically, we use `ip link add link <parent_port_name> name <subif_name> type vlan id <vlan_id>` to create a host sub port interface.
+This command implies the dependancy that a parent host interface must be created before the creation of a host sub port interface.
+At creation, a host sub port interface is always set to admin status up.
+
+Example:
+```
+ip link add link Ethernet64 name Ethernet64.10 type vlan id 10
```

Comment:
admin status of Ethernet64.10 is coupled with Ethernet64.
If Ethernet64.10 and Ethernet64 are both shutdown, startup Ethernet64 will also startup Ethernet64.10.
Is it better as follows:
```
ip link add br-Ethernet64 up type bridge
ip link add link br-Ethernet64 name Ethernet64.10 type vlan id 10
```

Reply:
I think it is ok to be coupled with Ethernet64 considering how it is configured in H/W. This is a sub-interface of Ethernet64 with a vlan tag and with above commands it is not related to Ethernet64. Not sure how the packet processing happens if we have above configuration and packet comes on Ethernet64 tagged with vlan 10.

Reply:
Since we have an explicit admin_status config for sub_intf, it is better to not to inherit the admin status from parent intf, if user has not configured the admin_status, we should use the default admin_status and when the packet comes on parent intf with VLAN 10 and the corresponding sub_intf created for VLAN 10 is admin down, kernel is expected to drop the packet which is expected, but having admin_status on sub_intf is of no significance, I think, we can always inherit admin_status from parent intf as Prince mentioned.